analytic_engine/src/instance/read.rs

```diff
@@ -331,72 +326,109 @@ impl Instance {
 
         read_view_by_time.into_values().collect()
     }
+
+    fn make_iter_options(&self, num_rows_per_row_group: usize) -> IterOptions {
+        self.iter_options.clone().unwrap_or(IterOptions {
+            batch_size: num_rows_per_row_group,
+        })
+    }
+}
+
+struct StreamStateOnMultiIters<I> {
+    iters: Vec<I>,
+    curr_iter_idx: usize,
+    projected_schema: ProjectedSchema,
 }
 
-// TODO(xikai): this is a hack way to implement SendableRecordBatchStream for
-// MergeIterator.
-fn iters_to_stream<T>(
-    collection: T,
-    runtime: &Runtime,
-    schema: &ProjectedSchema,
-) -> SendableRecordBatchStream
-where
-    T: IntoIterator + Send + 'static,
-    T::Item: RecordBatchWithKeyIterator,
-    T::IntoIter: Send,
-{
-    let (tx, rx) = mpsc::channel(RECORD_BATCH_READ_BUF_SIZE);
-    let projected_schema = schema.clone();
-
-    runtime.spawn(async move {
-        for mut iter in collection {
-            while let Some(record_batch) = iter.next_batch().await.transpose() {
-                let record_batch = record_batch.box_err().context(ErrWithSource {
-                    msg: "read record batch",
-                });
-
-                // Apply the projection to RecordBatchWithKey and gets the final RecordBatch.
-                let record_batch = record_batch.and_then(|batch_with_key| {
-                    // TODO(yingwen): Try to use projector to do this, which precompute row
-                    // indexes to project.
-                    batch_with_key
-                        .try_project(&projected_schema)
-                        .box_err()
-                        .context(ErrWithSource {
-                            msg: "project record batch",
-                        })
-                });
-
-                trace!("send next record batch:{:?}", record_batch);
-                if tx.send(record_batch).await.is_err() {
-                    error!("Failed to send record batch from the merge iterator");
-                    break;
-                }
+impl<I: RecordBatchWithKeyIterator + 'static> StreamStateOnMultiIters<I> {
+    fn is_exhausted(&self) -> bool {
+        self.curr_iter_idx >= self.iters.len()
+    }
+
+    fn advance(&mut self) {
+        self.curr_iter_idx += 1;
+    }
+
+    fn curr_iter_mut(&mut self) -> &mut I {
+        &mut self.iters[self.curr_iter_idx]
+    }
+
+    async fn fetch_next_batch(
+        &mut self,
+    ) -> Option<std::result::Result<RecordBatchWithKey, I::Error>> {
+        loop {
+            if self.is_exhausted() {
+                return None;
+            }
+
+            let iter = self.curr_iter_mut();
+            if let Some(v) = iter.next_batch().await.transpose() {
+                return Some(v);
             }
+
+            self.advance();
         }
+    }
+}
+
+fn iters_to_stream(
```

Comment:
In your new implementation, no channel is used as buffer, does this will slow down query when iters are among multiple sst segments?

Reply:
The channel here was not used as buffer, but used as a hack way to implement stream. Current implementation will perform better because the overhead caused by the channel is avoided.